job-runner/chain/traverser.go

```diff
@@ -81,209 +107,159 @@ func (f *traverserFactory) Make(jobChain proto.JobChain) (Traverser, error) {
-
-	NEXT_JOB:
-		for _, nextJ := range t.chain.NextJobs(doneJob.Id) {
-			nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJ.Id})
-
-			// Check to make sure the job is ready to run. It might not be if it has
-			// upstream dependencies that are still running.
-			if !t.chain.JobIsReady(nextJ.Id) {
-				nextJLogger.Infof("next job is not ready to run - not enqueuing it")
-				continue NEXT_JOB
-			}
-
-			nextJLogger.Infof("next job is ready to run - enqueuing it")
-
-			// Copy the jobData from the job that just finished to the next job.
-			// It is important to note that when a job has multiple parent
-			// jobs, it will get its jobData from whichever parent finishes
-			// last. Therefore, a job should never rely on jobData that was
-			// created during an unrelated sequence at any time earlier in
-			// the chain.
-			for k, v := range doneJob.Data {
-				nextJ.Data[k] = v
+	// Add the first jobs to the runJobChan.
+	// Prevent Stop()/shutdown() from closing runJobChan and don't enqueue jobs
+	// if the traverser has already been stopped or shutdown.
+	t.stopMux.Lock()
+	select {
+	case <-t.stopChan:
+		return
+	case <-t.shutdownChan:
+		return
+	default:
+		for _, job := range jobsToRun {
+			t.logger.Infof("sending initial job (%s) to runJobChan", job.Id)
+			if t.chain.IsSequenceStartJob(job.Id) {
+				// Starting a sequence, so increment sequence try count.
+				t.chain.IncrementSequenceTries(job.Id)
+				seqLogger := t.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": t.chain.SequenceTries(job.Id)})
+				seqLogger.Info("starting try of sequence")
 			}
-
-			t.runJobChan <- nextJ // add the job to the run queue
+			t.runJobChan <- job
 		}
+	}
+	t.stopMux.Unlock()
 
-		// Update chain repo for jobs next jobs just started ^
-		t.chainRepo.Set(t.chain)
+	// Reap jobs as they finish running. Stop reaping jobs when the doneJobChan
+	// is closed by runJobs - when the chain is done, stopped, or suspended.
+	for doneJob := range t.doneJobChan {
+		// Don't allow closing runJobChan / switching the reaper while we are
+		// reaping a job. Reap() should return quickly, so this is ok.
+		t.stopMux.Lock()
+		t.reaper.Reap(doneJob)
+		t.stopMux.Unlock()
 	}
 
-	return nil
+	// Indicate that the chain is done running. Stops waitForShutdown() listener.
+	select {
+	case <-t.doneChan:
+	default:
+		close(t.doneChan)
+	}
 }
 
-// Stop stops the traverser if it's running.
+// Stop stops a running chain by stopping all its running jobs.
 func (t *traverser) Stop() error {
+	// LOCK - can't close these chans while a job is being Reaped / initial
+	// jobs are being sent to runJobChan.
+	t.stopMux.Lock()
+
+	// Don't do anything if traverser has already been stopped.
+	select {
+	case <-t.stopChan:
+		return nil
+	default:
+	}
+
 	t.logger.Infof("stopping traverser and all jobs")
 
-	// Stop the traverser (i.e., stop running new jobs).
-	close(t.stopChan)
+	t.reaper = t.reaperFactory.MakeStopped()
```

Comment:
This needs a code comment. It's an important change and not obvious why we do it.